### PR TITLE
fix(Core/MySQL): _Query m_mStmt==null Crash

### PR DESCRIPTION
**Pull request by PkllonG**
## Changes Proposed:

This PR proposes changes to:
-  [ ] Core (units, players, creatures, game systems).
-  [ ] Scripts (bosses, spell scripts, creature scripts).
-  [ ] Database (SAI, creatures, etc).

## Issues Addressed:

- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/18741

## SOURCE:

The changes have been validated through:
- [ ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:

This PR has been:
- [ ] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

- [ ] This pull request can be tested by following the reproduction steps provided in the linked issue
- [ ] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

1.
2.
3.

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by PkllonG** [comment]
@pangolp Can you tell us what you think?

**Comment by kissingers** [comment]
Can add a log if null, so can know what sql problem?

**Comment by pangolp** [comment]
> @pangolp Can you tell us what you think?

As I mentioned in the pull request that I made, I really want the problem to be solved. It really makes no difference to me whether they approve of your method or mine. As long as the fix arrives and the error does not continue, I am satisfied.

**Comment by pangolp** [comment]
> Can add a log if null, so can know what sql problem?

This relationship is a problem, which causes the emulator to crash.

**Comment by PkllonG** [comment]
```2024-04-14 12:41:26 Could not fetch prepared statement 65 on database `characters`, connection type: synchronous.```
There is this

**Comment by PkllonG** [comment]
Player characters can be deleted normally,
I've observed that in the database, the mail that crashed the server was also deleted normally.

**Comment by PkllonG** [comment]
Can you help me if you need to add logs here? I don't speak English very well.

**Comment by kissingers** [comment]
> Can you help me if you need to add logs here? I don't speak English very well. @kissingers

It's horrible, my English is even worse and I totally rely on my browser's automatic translation and deepl

**Comment by PkllonG** [comment]
Can you help me @pangolp

**Comment by PkllonG** [comment]
> It doesn't crash after PR There is this `` 2024-04-14 12:41:26 Could not fetch prepared statement 65 on database `characters`, connection type: synchronous. ``

image
Hints have already been given here

**Comment by PkllonG** [comment]
I'm not crashing after testing

**Comment by pangolp** [comment]
The doubt I have is. Should both pull requests be applied or just one? Because this pull request is fine, however, we must tell the player why the player cannot be removed. In fact, we are debating that the message that we are giving today in the other PR does not have enough information for the player to understand what the cause of the problem is, and correct it before being able to continue with the elimination. In fact, you suggested the message that appears when changing faction, and I changed it, because at first, I also thought about it. But I checked the DELETE CHARACTER messages and was guided by the one I chose in the first instance. I explained the problem, so that we can all determine the best solution and we can solve the problem.

**Comment by PkllonG** [comment]
This PR should also prevent server crashes caused by other reasons

**Comment by pangolp** [comment]
> This PR should also prevent server crashes caused by other reasons

Did that mean we should add both changes or just one?

**Comment by PkllonG** [comment]
> > This PR should also prevent server crashes caused by other reasons
> 
> Did that mean we should add both changes or just one?

I recommend one. Because it doesn't crash, and it allows the player to delete the character normally

**Comment by PkllonG** [comment]
> > > This PR should also prevent server crashes caused by other reasons
> > 
> > 
> > Did that mean we should add both changes or just one?
> 
> I recommend one. Because it doesn't crash, and it allows the player to delete the character normally

The main thing is that it doesn't give players vague error prompts

**Comment by PkllonG** [comment]
[c143330](https://github.com/azerothcore/azerothcore-wotlk/pull/18744/commits/c143330fe2057fe486e5eaa95e514ae6e01a2699) We don't need to pop up the duplicate error log again.

File: src/server/database/Database/MySQLConnection.cpp [hunk 1]
```diff
@@ -218,7 +218,9 @@ bool MySQLConnection::Execute(PreparedStatementBase* stmt)
     uint32 index = stmt->GetIndex();
 
     MySQLPreparedStatement* m_mStmt = GetPreparedStatement(index);
-    ASSERT(m_mStmt); // Can only be null if preparation failed, server side error or bad query
+    //ASSERT(m_mStmt); // Can only be null if preparation failed, server side error or bad query
+    if (!m_mStmt)
+        return false;
```
Comment (FrancescoBorzi) [inline src/server/database/Database/MySQLConnection.cpp:223]:
```suggestion
    if (!m_mStmt)
    {
        return false;
    }
```

File: src/server/database/Database/MySQLConnection.cpp [hunk 2]
```diff
@@ -269,7 +271,9 @@ bool MySQLConnection::_Query(PreparedStatementBase* stmt, MySQLPreparedStatement
     uint32 index = stmt->GetIndex();
 
     MySQLPreparedStatement* m_mStmt = GetPreparedStatement(index);
-    ASSERT(m_mStmt);            // Can only be null if preparation failed, server side error or bad query
+    //ASSERT(m_mStmt);            // Can only be null if preparation failed, server side error or bad query
+    if (!m_mStmt)
+        return false;
```
Comment (FrancescoBorzi) [inline src/server/database/Database/MySQLConnection.cpp:276]:
```suggestion
    if (!m_mStmt)
    {
        return false;
    }
```

**Comment by kissingers** [comment]
This is a good fix, because even I use eluna to change some functions, when I write a wrong lua database insert, also will lead server crash,  if can avoid these database crash, and give the obvious mention, the server will be more robust.

**Comment by pangolp** [comment]
Link the solution to the issue. So that in case of a merge, it closes automatically

File: src/server/database/Database/MySQLConnection.cpp (hunk 1 above)
Comment (walkline) [reply src/server/database/Database/MySQLConnection.cpp]: @FrancescoBorzi you should check our latest code style convention. The first one was correct.

**Comment by walkline** [comment]
I don't like this solution. You shouldn't have a situation with a nil prepared statement, and you don't need to handle this case.

**Comment by pangolp** [comment]
> I don't like this solution. You shouldn't have a situation with a nil prepared statement, and you don't need to handle this case.

You have another alternative, linked in the topic. And if you don't like any of the two, you can open a third. We have to give you some solution. Because the greats do not have this fault. Even TC has this problem in its core 335 at least (in the rest of the nose). But, the servers, which have been online for years, magically, are all immune... because several people, upon seeing the problem, tried to see if they were also victims and found that they are not. They are an elite, and we are the scum.

**Comment by Nyeriah** [comment]
Thanks for the PR, but I don`t think this is an acceptable solution. If a statement fails rather have it crash than compromise DB data, thus the assert

**Pull request closed without merging**